ocaml/typing/jkind_types.mli

```diff
@@ -126,6 +126,7 @@ type 'type_expr history =
         rhs_history : 'type_expr history
       }
   | Creation of Jkind_intf.History.creation_reason
+  | Warning of 'type_expr history
```

Comment:
Though I know this is the design I described, I don't like it now that I see it implemented.

* The `update_reason` function is painful, having to work under `Warning`
* We never really care about `Warning`s in the middle of a tree.

So I think it would be better to have (leaving out the `'type_expr` parameter)

```ocaml
type history = (* just as it was *)
type t =
  { jkind : Jkind_desc.t;
    history : history;
    has_warned : bool
  }
```

Separating out the warning info -- and making it clear that it's just a bool, not (say) a count of how many times we've warned -- seems better.

Sorry to request a reimplementation here! Designs are sometimes hard to get right.

Reply:
Done.